controllers/storagecluster/provider_server.go

```diff
@@ -0,0 +1,264 @@
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+
+	"go.uber.org/multierr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+)
+
+var (
+	ServerImage = GetEnvVarOrPanic("PROVIDER_API_SERVER_IMAGE")
+)
+
+const (
+	OcsProviderServerName = "ocs-provider-server"
+)
+
+type ocsProviderServer struct{}
+
+func (o *ocsProviderServer) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !instance.Spec.AllowRemoteStorageConsumers {
+		r.Log.Info("nigoyal Spec.AllowRemoteStorageConsumers is disabled")
+		return o.ensureDeleted(r, instance)
+	}
+
+	r.Log.Info("nigoyal Spec.AllowRemoteStorageConsumers is enabled creating resources")
+
+	var finalErr error
+
+	for _, f := range []func(*StorageClusterReconciler, *ocsv1.StorageCluster) error{
+		o.createSecret,
+		o.createService,
+		o.createDeployment,
+	} {
+		err := f(r, instance)
+		if err != nil {
+			multierr.AppendInto(&finalErr, err)
+			continue
+		}
+	}
+
+	return finalErr
+}
+
+func (o *ocsProviderServer) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	// We do not check instance.Spec.AllowRemoteStorageConsumers because provider can disable this functionality
+	// and we need to delete the resources even the flag is not enabled (uninstall case).
+
+	// This func is directly called by the ensureCreated if the flag is disabled and deletes the resource
+	// Which means we do not need to call ensureDeleted while reconciling unless we are uninstalling
+
+	// NOTE: Do not add the check
+
+	var finalErr error
+
+	for _, resource := range []client.Object{
+		GetProviderApiServerSecret(instance),
+		GetProviderApiServerService(instance),
+		GetProviderApiServerDeployment(instance),
+	} {
+		err := r.Client.Delete(context.TODO(), resource)
+
+		if err != nil && !errors.IsNotFound(err) {
+			r.Log.Error(err, "nigoyal Failed to delete resource", "Kind", resource.GetObjectKind(), "Name", resource.GetName())
+			multierr.AppendInto(&finalErr, err)
+		}
+	}
+
+	if finalErr == nil {
+		r.Log.Info("nigoyal Resource deletion for provider succeeded")
+	}
+
+	return finalErr
+}
+
+func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredDeployment := GetProviderApiServerDeployment(instance)
+	actualDeployment := &appsv1.Deployment{}
+
+	actualDeployment.ObjectMeta = desiredDeployment.ObjectMeta
+	_, err := controllerutil.CreateOrUpdate(
+		context.TODO(), r.Client, actualDeployment,
+		func() error {
+			actualDeployment.Spec = desiredDeployment.Spec
+			return controllerutil.SetOwnerReference(instance, actualDeployment, r.Client.Scheme())
+		},
+	)
```

Comment:
I'm happy with this! For me this resolves this discussion below: https://github.com/red-hat-storage/ocs-operator/pull/1421/commits/bb44cbf3b72380f6aa1013d1c1d2834155e05d1e#r765682700

Reply:
Copying metadata is very problematic it removes additional labels and annotations that are set from the outside and this is something that we encounter again and again!

The metadata is both input and output we should never treat it as output-only or input only section

Reply:
To clarify, I am referring to line 94

Reply:
I agree, resolved the problem at all places.